src/H5L.c

```diff
@@ -503,45 +467,41 @@ H5L__create_hard_api_common(hid_t cur_loc_id, const char *cur_name, hid_t link_l
 
         /* Check if both objects are associated with the same VOL connector */
-        if (H5VL_cmp_connector_cls(&same_connector, curr_vol_obj->connector->cls,
-                                   link_vol_obj->connector->cls) < 0)
-            HGOTO_ERROR(H5E_FILE, H5E_CANTCOMPARE, FAIL, "can't compare connector classes");
-        if (same_connector)
+        if ((same_connector = H5VL_conn_same_class(H5VL_OBJ_CONNECTOR(curr_vol_obj),
+                                                   H5VL_OBJ_CONNECTOR(link_vol_obj))) < 0)
+            HGOTO_ERROR(H5E_LINK, H5E_CANTCOMPARE, FAIL, "can't compare connector classes");
+        if (!same_connector)
             HGOTO_ERROR(H5E_ARGS, H5E_BADTYPE, FAIL,
                         "Objects are accessed through different VOL connectors and can't be linked");
     } /* end if */
 
-    /* Construct a temporary VOL object */
-    if (curr_vol_obj)
-        (*tmp_vol_obj_ptr_ptr)->connector = curr_vol_obj->connector;
-    else {
-        if (NULL == link_vol_obj)
-            HGOTO_ERROR(H5E_LINK, H5E_BADVALUE, FAIL, "NULL VOL object pointer");
-
-        (*tmp_vol_obj_ptr_ptr)->connector = link_vol_obj->connector;
-    } /* end else */
-    if (link_vol_obj)
-        (*tmp_vol_obj_ptr_ptr)->data = link_vol_obj->data;
-    else
-        (*tmp_vol_obj_ptr_ptr)->data = NULL;
+    /* Set up new location struct */
+    link_loc_params.type     = H5VL_OBJECT_BY_NAME;
+    link_loc_params.obj_type = (link_vol_obj ? H5I_get_type(link_loc_id) : H5I_get_type(cur_loc_id));
+    link_loc_params.loc_data.loc_by_name.name    = link_name;
+    link_loc_params.loc_data.loc_by_name.lapl_id = lapl_id;
 
     /* Set up VOL callback arguments */
     vol_cb_args.op_type                        = H5VL_LINK_CREATE_HARD;
-    vol_cb_args.args.hard.curr_obj             = (curr_vol_obj ? curr_vol_obj->data : NULL);
+    vol_cb_args.args.hard.curr_obj             = (curr_vol_obj ? H5VL_OBJ_DATA(curr_vol_obj) : NULL);
     vol_cb_args.args.hard.curr_loc_params.type = H5VL_OBJECT_BY_NAME;
     vol_cb_args.args.hard.curr_loc_params.obj_type =
         (H5L_SAME_LOC != cur_loc_id ? H5I_get_type(cur_loc_id) : H5I_BADID);
     vol_cb_args.args.hard.curr_loc_params.loc_data.loc_by_name.name    = cur_name;
     vol_cb_args.args.hard.curr_loc_params.loc_data.loc_by_name.lapl_id = lapl_id;
 
     /* Create the link */
-    if (H5VL_link_create(&vol_cb_args, *tmp_vol_obj_ptr_ptr, &link_loc_params, lcpl_id, lapl_id,
-                         H5P_DATASET_XFER_DEFAULT, token_ptr) < 0)
+    if (H5VL_link_create(&vol_cb_args, (link_vol_obj ? link_vol_obj : curr_vol_obj), &link_loc_params,
+                         lcpl_id, lapl_id, H5P_DATASET_XFER_DEFAULT, token_ptr) < 0)
         HGOTO_ERROR(H5E_LINK, H5E_CANTCREATE, FAIL, "unable to create hard link");
 
+    /* Set the connector to use for async operations */
+    if (connector)
+        *connector = (link_vol_obj ? H5VL_OBJ_CONNECTOR(link_vol_obj) : H5VL_OBJ_CONNECTOR(curr_vol_obj));
```

Comment:
Since this returns a pointer to the connector, I think it should increment the ref count and require the caller to release it after its done.